### PR TITLE
[RFC] Add workflow to use reviewdog

### PR DESCRIPTION
**Pull request by himkt**
This PR introduces a new workflow for reviewdog's [action-suggester](https://github.com/reviewdog/action-suggester).

## Motivation
Automate the process of code formatting.

I verified the behavior on my repo.
https://github.com/himkt/optuna-black-suggester/pull/4

Screen Shot 2022-03-03 at 22 32 35

### CODE REVIEW

**Comment by nzw0301** [comment]
How about `blackdoc` and `isort` that also fix the format.

File: .github/workflows/reviewdog.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    - name: Blackdoc
+      run: blackdoc .
+
+    - name: Isort
+      run: isort .
+
+    - name: Reviewdog
+      uses: reviewdog/action-suggester@v1
+      with:
+        tool_name: black
```
Comment (nzw0301) [inline .github/workflows/reviewdog.yml:34]: Should we change this value as well?

Comment (himkt) [reply .github/workflows/reviewdog.yml]: 👍 

Before: https://github.com/himkt/optuna-black-suggester/pull/5
After: https://github.com/himkt/optuna-black-suggester/pull/6

Comment (nzw0301) [reply .github/workflows/reviewdog.yml]: lovely.

File: .github/workflows/reviewdog.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+name: reviewdog-suggester
+
+on:
+  pull_request: {}
+
+jobs:
+  black:
```
Comment (nzw0301) [inline .github/workflows/reviewdog.yml:7]: Sorry, could you rename this?

Comment (himkt) [reply .github/workflows/reviewdog.yml]: So sorry for bothering you... [c48d396](https://github.com/optuna/optuna/pull/3357/commits/c48d396645557f610e602f645004a4d077b305f0) fixed.

**Comment by HideakiImamura** [comment]
Let me ask a simple question. Is this workflow is triggered on each push on the PR?

**Comment by nzw0301** [comment]
> Let me ask a simple question. Is this workflow is triggered on each push on the PR?

I think so. So it might be showing so many comments :(

**Comment by nzw0301** [comment]
Can we hide the previous comment by reviewdog when reviewdog is triggered by a new commit to a pull request?

**Comment by himkt** [comment]
Sorry for the late.

> Is this workflow is triggered on each push on the PR?

You're right. Now the suggestion is triggered on each push event.
https://github.com/himkt/optuna-black-suggester/pull/7

> Can we hide the previous comment by reviewdog when reviewdog is triggered by a new commit to a pull request?

I think it's potentially possible but reviewdog/action-suggester doesn't provide the such feature.
It may need to implement our own action job using GitHub API. I'm not sure if it'd be a good idea to create/maintain the such thing.

I think there are two options:
- stop introducing this suggestion job
- only trigger the job on the creation of PR

**Comment by HideakiImamura** [comment]
Thanks for the suggestion. As another suggestion, how about triggering the job by some comment?

**Comment by himkt** [comment]
Now the workflow is modified as follows:

- workflow is triggered on event: `pull_request` && action: `opened`. This means reviewdog checks a codebase when user creates a pull request
- workflow is triggered on event: `pull_request` && action: `review_request`. Reviewdog checks a codebase when user requests the specific `team` for a review. You can see the example in https://github.com/himkt-org/optuna-black-suggester/pull/4

The reason why I don't make a change to trigger workflow on a comment is that a comment on a PR is not an event of `pull_request` but of `issue_comment`. action-suggester only supports a `pull_request` event (🔗 https://github.com/reviewdog/action-suggester#usage-example).

To enable the workflow in Optuna, we have to create a team `reviewdog` in Optuna org or optuna/optuna (it is OK that no member is in the group) for the target of the trigger.

File: .github/workflows/reviewdog.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+on:
+  pull_request:
+    types: [opened, review_requested]
+
+jobs:
+  reviewdog:
+    runs-on: ubuntu-latest
+
+    if: github.event.action == 'opened' || github.event.requested_team.name == 'reviewdog'
```
Comment (himkt) [inline .github/workflows/reviewdog.yml:11]: Although I couldn't find `requested_team` in the payload documentation, but it actually exists and it is documented in the actions reference.

- payload: https://docs.github.com/en/rest/reference/pulls#get-a-pull-request
- actions document: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

**Comment by himkt** [comment]
@HideakiImamura @nzw0301 PTAL again. :bow:

File: .github/workflows/reviewdog.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    - uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
+
+    - name: Install
+      run: |
+        python -m pip install -U pip
+        pip install --progress-bar off -U .[checking]
+
+    - name: Black
```
[the remark this answers is not in the document]
Comment (himkt) [reply .github/workflows/reviewdog.yml]: :bow:

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update and sorry for the late. LGTM!

**Review by nzw0301 — APPROVED** [review]
LGTM! This would be helpful for PR reviewers.

**Pull request merged**